### PR TITLE
Update breakpoints window position and always open expression evaluator

### PR DESCRIPTION
**Pull request by jtulach**
I am using the Apache NetBeans VSCode extension frequently and one thing I found very useful is the location of breakpoints view in the VSCode default layout. It is not in the _Output_ area, but it is on the left. It is very convenient to have the breakpoints view always visible when one wants to quickly enable and disable certain breakpoints. I'd like the users of classical NetBeans UI to have similar experience. This is how the new default layout looks like when one hits a Java breakpoint:

Breakpoints view

The breakpoints view shows over navigator. I believe that's OK as in 90% of times I am not using navigator when debugging. User can of course drag the breakpoints view to another mode easily. Or just press Ctrl-7 to bring the navigator to the top of the mode.

### CODE REVIEW

**Comment by jtulach** [comment]
Please compare the NetBeans UI look with VSCode extension:

VSCode Breakpoins

Please find _breakpoints view_ in the bottom left corner. Visible by default, easy to click and _enable/disable a breakpoint_.

**Comment by mbien** [comment]
nothing really against it but I had on my todo list to check if the "Evaluate Expression" window could be opened by default and I would have put it to the same spot the breakpoints windows is in the screenshot (it interacts with the variables window, so it makes sense to have it side by side).
This feature is currently a bit hidden.

image

But I suppose this doesn't really conflict since it would simply add another tab to the view.

**Comment by jtulach** [comment]
> .. if the "Evaluate Expression" window could be opened by default and I would have put it to the same spot the breakpoints window is in the screenshot above (it interacts with the variables window, so it makes sense to have it side by side). 

I wanted to start with something simple and small, but yes, repositioning "Evaluate Expression" deserves to be done as well. This is the layout I use in my IDE:

Breakpoints & evaluate expression views

I can modify the PR to reposition "Evaluate Expression" window as well, if there is a consensus where to put it. The current location (below editor and above output) is needlessly wide and takes away precious space of editor or variables view.

**Comment by mbien** [comment]
> The current location (below editor and above output) is needlessly wide and takes away precious space of editor or variables view.

yes I agree. How about moving it to the lowest position on the left column like on my screenshot (breakpoints would be above it)? This would make it easier for users to resize the other two debugging views and it would be right next to the variables view.

**Comment by jtulach** [comment]
After 98f14f8 the default layout looks like this:

image

> to the lowest position on the left column like on my screenshot (breakpoints would be above it)? 

Oh, UI. Hard to decide what's better. I like it my way, of course.

Anyone else whats to vote? Should it be (from top to bottom):
- debugging view, evaluate expression, breakpoints (vote +1)
- debugging view, breakpoints, evaluate expression (vote -1)

**Comment by mbien** [comment]
in case this won't get enough feedback this might be a good topic for the dev list. We had some "good defaults" discussions before: [round 1](https://lists.apache.org/thread/g9w0038rdsksony6xyl4q42vnj4jrjk4) and [round 2](https://lists.apache.org/thread/rdflfdbt2v0cxz9mk198tonm0hnff1bg) but not recently

**Comment by MartinBalin** [comment]
Getting used to VSCode UI as well I find this useful and good UX.

**Comment by geertjanw** [comment]
I’m in favor, go for it. If not enough reactions, publish/promote this via Foojay.io (bit.ly/join-foojay-slack).

**Comment by mbien** [comment]
i ran this PR on a fresh user dir and the eval window did not open automatically during debugging, could this be added?

I found also an unrelated issue: `window -> reset windows` does seem to only restore the UI state of the "normal mode", if you close breakpoints or debugging view they won't reset. (also don't reset windows _while_ debugging or it will make the "debug mode" look like the "normal mode")

File: ide/spi.debugger.ui/src/org/netbeans/modules/debugger/resources/mf-layer.xml [hunk 1]
```diff
@@ -330,6 +330,11 @@
         </folder>
 
         <folder name="Modes">
+            <folder name="navigator">
+                <file name="breakpointsView.wstcref" url="modes/navigator/breakpoints.wstcref">
+                    <attr name="position" intvalue="400"/>
+                </file>
+            </folder>
```
Comment (mbien) [inline ide/spi.debugger.ui/src/org/netbeans/modules/debugger/resources/mf-layer.xml:337]: nitpick: not sure if the registration order is relevant, but consider moving this down below the evaluator folder, so that the registration order resembles somewhat the order of appearance - makes it easier to read for me at least.

**Comment by jtulach** [comment]
> Lets ignore this issue for now - but lets try to open the eval window by default at least.

When I ran with empty user dir, the evaluation window got opened on first debug session. I'll verify before integration again.

Other than that I see six thumbs up for

> debugging view, evaluate expression, breakpoints

so, let's leave the layout as it is for now.

**Comment by mbien** [comment]
> Other than that I see six thumbs up for

yes my thumb too :)


> When I ran with empty user dir, the evaluation window got opened on first debug session. I'll verify before integration again.

how it looks for me, I only opened the project, set the breakpoint and hit debug:
image

**Comment by jtulach** [comment]
> > When I ran with empty user dir, the evaluation window got opened on first debug session. I'll verify before integration again.
> 
> how it looks for me, I only opened the project, set the breakpoint and hit debug:

Fixed since 14fcf69 - now the debugger forcibly opens `evaluatorPane` in its _session added_ code.

File: ide/spi.debugger.ui/src/org/netbeans/modules/debugger/resources/mf-layer.xml (hunk 1 above)
Comment (jtulach) [reply ide/spi.debugger.ui/src/org/netbeans/modules/debugger/resources/mf-layer.xml]: The order is completely irrelevant. There is even `output` between the folders. Anyway, it is no harm to 9b461e1335

**Comment by mbien** [comment]
aweseome, tested again and it worked. And thanks for moving the view registration around, I think its less confusing when things which belong together are declared in the same general area.

Please don't forget to squash before merge.

**Comment by jtulach** [comment]
@mbien, why do you think the _commits needs squashing_? I see:
- [Initial position of breakpoints is next to navigator](https://github.com/apache/netbeans/pull/7496/commits/18db1f1c186a9d32176ad94c881f58393c21df86)
- [Put Evaluate Expression between debugger view and breakpoints](https://github.com/apache/netbeans/pull/7496/commits/98f14f8d2c3b98ad10c275dd34886af8aa5f126f)
- [Make sure evaluatorPane is opened when debugging starts](https://github.com/apache/netbeans/pull/7496/commits/14fcf6990630ef2a1c5d25bc6e132a0eadb65ce1)
- [Moving the navigator definition below other ones](https://github.com/apache/netbeans/pull/7496/commits/9b461e13359e599cbb607e06f1ee5198f195b55b)

Maybe the forth one is slightly unnecessary, but the _first three clearly describe_ what's the intention of the change. I prefer to keep them - e.g. merge, not squash.

**Comment by mbien** [comment]
@jtulach well, the original intention why we allowed multiple commits per PR was to not discourage cleanups in bugfix PRs for example. One commit could refactor, the other fix the bug/add a feature without obfuscating what is happening. (main reason why I advocated for the option of multi commit PRs at least)

But simple PRs like here can be described by a few bullet points in the commit msg and everyone can still see what is happening while looking at the diff.

Many projects don't allow merges at all, they auto squash+rebase on "merge" (which also leads to an easy to understand history). The reason we decided to disable that one is because there were some problems with author headers in past. If the git account isn't properly setup mails can disappear etc. (I forgot the details but there were problems when author != committer in some circumstances). Sometimes too much choice is also bad. We had accidental merges where someone thought that squash was selected and the whole dev cycle got merged. Now there is only merge, so the last step of the review is to make sure the PR is properly squashed and that there are no surprises in the [commit header](https://patch-diff.githubusercontent.com/raw/apache/netbeans/pull/7496.patch). Keep it simple works often best.

Having multiple commits per PR can also cause problems during git bisect, since extra care is needed to make sure each individual commit builds and tests.

so in short: the thought is that most PRs can be one commit, but it can be useful to allow multiple for large PRs or special situations, e.g having commits from different authors in the same PR etc.

I try to follow the same rule of course: the last item I worked on ended up being 4 commits, split into 3 PRs: https://github.com/apache/netbeans/pull/7512, https://github.com/apache/netbeans/pull/7515, https://github.com/apache/netbeans/pull/7513 (I almost made that 3 commits though)

**Comment by jtulach** [comment]
NetBeans never preferred squashing over merges while I was actively around. 

> Many projects don't allow merges at all, they auto squash+rebase on "merge"

However NetBeans isn't one of such projects and I hope it will never be.

> simple PRs like here

This is not a simple PR! It was when I only proposed to change:

- [Initial position of breakpoints is next to navigator](https://github.com/apache/netbeans/pull/7496/commits/18db1f1c186a9d32176ad94c881f58393c21df86)

but then you asked me to do additional (and unrelated) change repositioning the _evaluate expression_ view:

- [Put Evaluate Expression between debugger view and breakpoints](https://github.com/apache/netbeans/pull/7496/commits/98f14f8d2c3b98ad10c275dd34886af8aa5f126f)

and yet another to open _evaluate expression view_ when debugging starts:

- [Make sure evaluatorPane is opened when debugging starts](https://github.com/apache/netbeans/pull/7496/commits/14fcf6990630ef2a1c5d25bc6e132a0eadb65ce1)


Those are three different changes and I refuse to squash them. Can you remove your _needs squashing_ label? If not, and there really is a policy we have to squash commits (link please), then tell me how can I proceed without squashing? Shall I drop the evaluate expression changes from this PR and integrate only [Initial position of breakpoints is next to navigator](https://github.com/apache/netbeans/pull/7496/commits/18db1f1c186a9d32176ad94c881f58393c21df86) and leave the rest for some other PRs?

**Comment by mbien** [comment]
> Shall I drop the evaluate expression changes from this PR and integrate only [Initial position of breakpoints is next to navigator](https://github.com/apache/netbeans/pull/7496/commits/18db1f1c186a9d32176ad94c881f58393c21df86) and leave the rest for some other PRs?

here is how I would integrate a change like this:

example 1:
 - rename this PR to "Update breakpoints window position and always open expression evaluator" (or something similar which mentions both user facing changes)
 - merge with 2 commits, one per feature
 
example 2:
 - rename as in 1
 - merge as 1 commit
 
example 3:
 - keep title
 - merge breakpoint window changes as 1 commit
 - move rest to other PR

I think all examples above would generate good results when read from the release notes and also from git history. In the example given previously in https://github.com/apache/netbeans/pull/7496#issuecomment-2197416124, i opted for 3 since it became a larger change, so i made it 3 PRs. But this diff is easy to parse, so example 2 or 1 would be completely fine IMO.

> but then you asked me to do additional (and unrelated) change repositioning the evaluate expression view:

I promise I won't do this again. But my intention was to not _ask_ you to do anything. I only mentioned that I had something on my todo list which slightly overlaps with this change, please read my initial post https://github.com/apache/netbeans/pull/7496#issuecomment-2179552776. Once you implemented the expression eval window positioning, I tested it and mentioned that it did not open. (that was also when I noticed https://github.com/apache/netbeans/pull/7528 and that the "Reset Windows" feature does not work as expected).

**Comment by jtulach** [comment]
> * rename this PR to "Update breakpoints window position and always open expression evaluator"
> * merge with 2 commits, one per feature

Renamed and force-pushed as:

- [Initial position of breakpoints view is next to navigator](https://github.com/apache/netbeans/pull/7496/commits/371383fe43c119ab9bfbe54024a26c35b0f66f4e)
- [Always open expression evaluator view](https://github.com/apache/netbeans/pull/7496/commits/66953b84c3236221dd7ebe9d9f7f0d1705f4cc68)

**Review by mbien — APPROVED** [review]
looks good!
